relayer/relayer.go

```diff
@@ -6,40 +6,39 @@ package relayer
 import (
 	"context"
 
-	"github.com/ChainSafe/sygma-core/types"
+	"github.com/ChainSafe/sygma-core/relayer/message"
+	"github.com/ChainSafe/sygma-core/relayer/proposal"
 	"github.com/rs/zerolog/log"
 )
 
-type DepositMeter interface {
-	TrackDepositMessage(m *types.Message)
-	TrackExecutionError(m *types.Message)
-	TrackSuccessfulExecutionLatency(m *types.Message)
-}
-
 type RelayedChain interface {
+	// PollEvents starts listening for on-chain events
 	PollEvents(ctx context.Context)
-	Write(messages []*types.Message) error
+	// ReceiveMessage accepts the message from the source chain and converts it into
+	// a Proposal to be submitted on-chain
+	ReceiveMessage(m *message.Message) (*proposal.Proposal, error)
+	// Write submits proposals on-chain.
+	// If multiple proposals submitted they are expected to be able to be batched.
+	Write(proposals []*proposal.Proposal) error
 	DomainID() uint8
 }
 
-func NewRelayer(chains []RelayedChain, metrics DepositMeter) *Relayer {
-	return &Relayer{relayedChains: chains, metrics: metrics}
+func NewRelayer(chains map[uint8]RelayedChain) *Relayer {
+	return &Relayer{relayedChains: chains}
 }
 
 type Relayer struct {
-	metrics       DepositMeter
-	relayedChains []RelayedChain
-	registry      map[uint8]RelayedChain
+	relayedChains map[uint8]RelayedChain
 }
 
-// Start function starts the relayer. Relayer routine is starting all the chains
-// and passing them with a channel that accepts unified cross chain message format
-func (r *Relayer) Start(ctx context.Context, msgChan chan []*types.Message) {
-	log.Debug().Msgf("Starting relayer")
+// Start function starts polling events for each chain and listens to cross-chain messages.
+// If an array of messages is sent to the channel they are expected to be to the same destination and
+// able to be handled in batches.
+func (r *Relayer) Start(ctx context.Context, msgChan chan []*message.Message) {
+	log.Info().Msgf("Starting relayer")
 
 	for _, c := range r.relayedChains {
 		log.Debug().Msgf("Starting chain %v", c.DomainID())
-		r.addRelayedChain(c)
 		go c.PollEvents(ctx)
```

Comment:
The channel is expected to be passed into message and event handlers and not into the listener directly.
Otherwise you cannot send messages back to the source